### PR TITLE
Pushdown aggregator group by through read hint

### PR DESCRIPTION
**Pull request by garrettlish**
Implement https://github.com/prometheus/prometheus/issues/6400

Signed-off-by: xiancli <xiancli@ebay.com>

### CODE REVIEW

File: prompb/types.proto [hunk 1]
```diff
@@ -56,6 +56,8 @@ message ReadHints {
   string func = 2;    // String representation of surrounding function or aggregation.
   int64 start_ms = 3; // Start time in milliseconds.
   int64 end_ms = 4;   // End time in milliseconds.
+  repeated string groups = 5; // List of label names of any fields used in a group-by aggregation
```
Comment (brian-brazil) [inline prompb/types.proto]: Full stops. And there's no such thing as a "group-by".

**Comment by brian-brazil** [comment]
Also we should think a bit about counters here I think, as common use cases look like sum(rate) or historgam_quantile(sum(rate())).

@beorn7 this aspect likely is of interest.

File: prompb/types.proto (hunk 1 above)
Comment (garrettlish) [reply prompb/types.proto]: sure, will change it to `group_by`

File: prompb/types.proto [hunk 2]
```diff
@@ -56,6 +56,8 @@ message ReadHints {
   string func = 2;    // String representation of surrounding function or aggregation.
   int64 start_ms = 3; // Start time in milliseconds.
   int64 end_ms = 4;   // End time in milliseconds.
+  repeated string group_by = 5; // List of label names of any fields used in a group-by aggregation
```
Comment (brian-brazil) [inline prompb/types.proto]: grouping is what we use internally, so probably best to stick with that.

File: prompb/types.pb.go [hunk 3]
```diff
@@ -430,6 +432,20 @@ func (m *ReadHints) GetEndMs() int64 {
...
+		return m.GroupBy
+	}
+	return nil
+}
+
+func (m *ReadHints) GetWithout() bool {
+	if m != nil {
+		return m.Without
+	}
+	return false
```
Comment (brian-brazil) [inline prompb/types.pb.go]: If missing this would mean `by()`, which is not the same as not having a grouping. True is the right default here.

**Comment by garrettlish** [comment]
@brian-brazil To support `sum(rate)`, besides the spatial aggregation read hint, we might need to introduce a temporal aggregation read hint and time window, something like `ta=rate:5m`. If so, maybe we can extend the PR to support pushdown TA functions as well (such as `<aggregation>_over_time`).

File: prompb/types.proto (hunk 2 above)
Comment (garrettlish) [reply prompb/types.proto]: sure

File: prompb/types.pb.go (hunk 3 above)
Comment (garrettlish) [reply prompb/types.pb.go]: will correct it

Comment (brian-brazil) [reply prompb/types.pb.go]: Thinking further, we actually want to change the field to By (rather than Without) and default to false. That way a Prometheus that doesn't support this looks the same as one that does support it but wants no grouping.

Comment (garrettlish) [reply prompb/types.pb.go]: Yes, agree, I did change it to `By` and default set to false, please see https://github.com/prometheus/prometheus/pull/6401/commits/95cb9b4ba4b627a903f044b4c8c8b090e609371b

File: prompb/types.proto (hunk 1 above)
Comment (brian-brazil) [reply prompb/types.proto]: `Grouping` is what we use elsewhere. Also s/group-by//, this isn't SQL.

File: storage/interface.go [hunk 4]
```diff
@@ -71,6 +71,9 @@ type SelectParams struct {
 
 	Step int64  // Query step size in milliseconds.
 	Func string // String representation of surrounding function or aggregation.
+
+	Groups []string // List of label names of any fields used in a group-by aggregation
+	By     bool     // Indicate whether it is without or by
```
Comment (beorn7) [inline storage/interface.go]: Nit: Please end comments with periods, like above.

**Comment by beorn7** [comment]
> @beorn7 this aspect likely is of interest.

I guess because of the upcoming histogram with 40,000 buckets? At the moment, I'd think that needs even deeper integration, but we'll see.

**Comment by brian-brazil** [comment]
> I guess because of the upcoming histogram with 40,000 buckets? At the moment, I'd think that needs even deeper integration, but we'll see.

Yes. In the most optimistic case something like this may turn out to be enough, without having to add layering violations to the interfaces themselves.

**Comment by brian-brazil** [comment]
> Also, does this PR supports without () or rather by () ?

Yes, set `By` to true.

**Comment by bwplotka** [comment]
>  > Also, does this PR supports without () or rather by () ?

> Yes, set By to true.

Ah yes all good. And for `without ()` we don't group as expected.

**Comment by brian-brazil** [comment]
> Ah yes all good.

It only took me two goes to figure it out :)

File: promql/engine_test.go [hunk 5]
```diff
@@ -1028,3 +1029,44 @@ func TestSubquerySelector(t *testing.T) {
...
+	by, groups = inspectExpr("max by (dim1) (avg_over_time(metric[5m]))", extractGroupsFromPath)
+	testutil.Assert(t, len(by) == 0 && len(groups) == 0, "Expected 0 - 0 but got %s - %s", nil, len(by), len(groups))
+
+	by, groups = inspectExpr("(max by (dim1) (metric))[5m:1m]", extractGroupsFromPath)
+	testutil.Assert(t, by[0] == true && reflect.DeepEqual(groups[0], expectedGroupBy), "Expected true - %s but got %t - %s", expectedGroupBy, by, groups)
+
+	by, groups = inspectExpr("(max without (dim1) (metric))[5m:1m]", extractGroupsFromPath)
+	testutil.Assert(t, by[0] == false && reflect.DeepEqual(groups[0], expectedGroupBy), "Expected false - %s but got %t - %s", expectedGroupBy, by, groups)
+
+	by, groups = inspectExpr("max by (dim1) (metric) - min by (dim2) (metric2)", extractGroupsFromPath)
```
[the remark this answers is not in the document]
Comment (garrettlish) [reply promql/engine_test.go]: @bwplotka we will only extract inner max function along with dim1 & dim2 grouping into read hint

File: promql/engine_test.go [hunk 6]
```diff
@@ -1028,3 +1029,44 @@ func TestSubquerySelector(t *testing.T) {
 		}
 	}
 }
+
+func inspectExpr(expr string, f func(p []Node) (bool, []string)) ([]bool, [][]string) {
```
[the remark this answers is not in the document]
Comment (garrettlish) [reply promql/engine_test.go]: make sense, we can using `extractGroupsFromPath` directly rather than a parameter, let me also try using `TestParamsSetCorrectly ` to test grouping parameters. Thanks @bwplotka

**Comment by garrettlish** [comment]
@bwplotka 
> How this will work exactly on sum(rate(.... [5m])? E.g What you will return for such select from some remote storage?

This PR supports spatial aggregation only, to support temporal aggregation, I think we need to introduce another temporal aggregation read hint and time window, something like `ta=rate:5m`. 

> I think the problem is that we might have to say stop at some point, like how to make sure to support queries like: sum(sum by(handler) (sum by (handler, instance, method) (metric))) which in theory is the same as sum(metric)?

This PR only handles the inner aggregations, we only pushdown sum function along with handler, instance, method grouping to read hint in your example.

> I think I like the idea of supporting common queries like sum(rate or histogram ones.

This PR doesn't support sum(rate) & histogram. To support them, I think we need to introduce another temporal aggregation read hint and time window, something like `ta=rate:5m`.

> Also, does this PR supports without () or rather by () ? I think we might miss this case. For this, we might need to add maybe Grouping type for in proto (?) to tell if grouping is actually used or not. Do you mind adding test cases for this?

Yes, we do support both without and by, for without unit test, please refer to https://github.com/prometheus/prometheus/pull/6401/files#diff-97f10a261de0bb743875fe9cc3a969d1R367-R373.

**Comment by garrettlish** [comment]
@brian-brazil @bwplotka @beorn7 comments fixed, besides, I also add temporal aggregation pushdown support, now this PR supports the following 3 scenarios but not support subqueries.
1. TA only, such as `rate(range_vector)`, `sum_over_time(range_vector)` - pushdown func name and time window
2. SA only, such as `sum by (dimensions) (instant_vector)` - pushdown func and group by dimensions
3. SA with TA, such `sum by (dimensions) (rate(range_vector))` - pushdown func, group by dimensions and time window

File: prompb/types.proto [hunk 7]
```diff
@@ -56,6 +56,9 @@ message ReadHints {
   string func = 2;    // String representation of surrounding function or aggregation.
   int64 start_ms = 3; // Start time in milliseconds.
   int64 end_ms = 4;   // End time in milliseconds.
+  repeated string grouping = 5; // List of label names of any fields used in aggregation.
+  bool by = 6; // Indicate whether it is without or by, it will be false if grouping is not used.
+  string ta = 7; // Temporal aggregation read hint include func name and time window, such as rate:5m0s.
```
Comment (brian-brazil) [inline prompb/types.proto]: No need for the function name, we just need the time and it should be in milliseconds like the other. The field name also doesn't go over the wire, so no need to abbreviate it. `range` is what we'd usually call this.

Comment (garrettlish) [reply prompb/types.proto]: @brian-brazil If we don't include function name, how can we distinguish `sum by (dim) (rate(range_vector))` and `avg by (dim) (rate(range_vector))`? `sum by (dim) (sum_over_time(range_vector))` and `max by (dim) (sum_over_time(range_vector))`?

Comment (brian-brazil) [reply prompb/types.proto]: The function name is already in `func`

Comment (garrettlish) [reply prompb/types.proto]: @brian-brazil In this PR, we'd like to support 1) temporal aggregation, 2) spatial aggregation 3) temporal aggregation. For 1) & 2), the `func` will be temporal aggregation func name (such as `rate`, `sum_over_time`, etc.) name or spatial aggregation func name (such as `sum`, `avg`, etc.). However, for 3), the `func` will be spatial aggregation func name, we don't have a place to record the temporal aggregation func name.
For example:
`sum by (dim) (rate(range_vector))`, if we use `rate` as `func`, then we cannot identify spatial aggregation name `sum`.

Comment (brian-brazil) [reply prompb/types.proto]: We could make func a list, but we're already quite bit of scope creep here so I'd say let's get this in without worrying about 3.

Comment (garrettlish) [reply prompb/types.proto]: make sense, let me make the change

File: prompb/types.proto [hunk 8]
```diff
@@ -56,6 +56,9 @@ message ReadHints {
   string func = 2;    // String representation of surrounding function or aggregation.
   int64 start_ms = 3; // Start time in milliseconds.
   int64 end_ms = 4;   // End time in milliseconds.
+  repeated string grouping = 5; // List of label names of any fields used in aggregation.
+  bool by = 6; // Indicate whether it is without or by, it will be false if grouping is not used.
+  int64 range_ms = 7; // Temporal aggregation time window in millisecond.
```
Comment (brian-brazil) [inline prompb/types.proto]: This is a bit wordy, and we don't use these terms in the docs. "Range vector selector range" covers it.

File: prompb/types.proto [hunk 9]
```diff
@@ -56,6 +56,9 @@ message ReadHints {
   string func = 2;    // String representation of surrounding function or aggregation.
   int64 start_ms = 3; // Start time in milliseconds.
   int64 end_ms = 4;   // End time in milliseconds.
+  repeated string grouping = 5; // List of label names of any fields used in aggregation.
+  bool by = 6; // Indicate whether it is without or by, it will be false if grouping is not used.
```
Comment (brian-brazil) [inline prompb/types.proto]: No need for anything after the comma.

Comment (garrettlish) [reply prompb/types.proto]: sure

File: prompb/types.proto [hunk 10]
```diff
@@ -56,6 +56,9 @@ message ReadHints {
   string func = 2;    // String representation of surrounding function or aggregation.
   int64 start_ms = 3; // Start time in milliseconds.
   int64 end_ms = 4;   // End time in milliseconds.
+  repeated string grouping = 5; // List of label names of any fields used in aggregation.
```
Comment (brian-brazil) [inline prompb/types.proto]: s/of any fields/

PromQL doesn't have fields.

File: prompb/types.proto [hunk 11]
```diff
@@ -56,6 +56,9 @@ message ReadHints {
...
+  bool by = 6; // Indicate whether it is without or by.
+  int64 range_ms = 7; // Range vector selector range in millisecond.
```
Comment (brian-brazil) [inline prompb/types.proto]: milliseconds

File: prompb/types.proto (hunk 10 above)
Comment (garrettlish) [reply prompb/types.proto]: What about `List of label names used in aggregation. `?

File: prompb/types.proto (hunk 11 above)
Comment (garrettlish) [reply prompb/types.proto]: sorry, let me correct it.

**Comment by brian-brazil** [comment]
Thanks!

**Comment by bwplotka** [comment]
:+1:

**Comment by sudeepknair** [comment]
👍
